### ISSUE TITLE
In USFM3, Quote marks need to be on the same line as the word they touch

### ISSUE THREAD

**Issue by benjore** [unfoldingWord/translationCore#5799]
### User Story

As a proofreader, I want the `"` marks to be on the same line as the word they modify (in the same milestone so that we have good, valid USFM.

Talk to Larry, Robert & Klappy to get all the specs.

**_Update_**
Also growing to include correction of format of `\zaln-s` and `\k-s` markers

### Features / Specifications

- [x] Cross Reference #5133
- [ ] Cross Reference https://github.com/unfoldingWord-dev/translationCore/issues/5800
- [ ] Cross Reference https://github.com/unfoldingWord-dev/translationCore/issues/5051 - see comment by Robert

### Definition of Done

- [x] add `\usfm 3.0` marker to header on export **[1.1.3 (5a14a95)The usfm 3.0 marker is being added to all usfm exports.]**
- [x] fix export so that extraneous new lines are not added (only in place of space before) before `\zaln` and `\w` and `\k` markers **[Looks good in 1.1.3 (982b0d9) ]**
- [x] maintain compatibility importing both with and without the new lines (using the `\usfm 3.0` marker as flag to expect new format)  **[Looks good in 1.1.3 (982b0d9) ]**
- [x] correct output format of `\zaln-s` markers (adding the `\*` terminator) **[Looks good in 1.1.3 (5a14a95)]**
- [x] maintain compatibility with importing both new and old format `\zaln-s` markers  **[Looks good in 1.1.3 (5a14a95)]**
- [x] correct output format of `\k-s` markers (adding the `\*` terminator) **[1.1.3 (5a14a95) I don't know anyway to test this]**
- [x] maintain compatibility with importing both new and old format `\k-s` markers

**Comment by lrsallee** [comment]
Dropping in sample files for ULT Galatians in original tC-exported form, and with manual corrections. Everything fixed except the splitting of "'s" from possessive words. This page would not accept .usfm files, so I changed the extensions to .txt. They need to be changed back for use.
[49-GAL_usfm2_corrected.txt](https://github.com/unfoldingWord-dev/translationCore/files/2857779/49-GAL_usfm2_corrected.txt)
[49-GAL_usfm2_original.txt](https://github.com/unfoldingWord-dev/translationCore/files/2857780/49-GAL_usfm2_original.txt)
[49-GAL_usfm3_original.txt](https://github.com/unfoldingWord-dev/translationCore/files/2857782/49-GAL_usfm3_original.txt)

**Comment by lrsallee** [comment]
Issues fixed:

- moved \s5 markers to new line (USFM2)
- moved \q markers to new line (USFM2)
- Added \usfm 3.0 identified to USFM# (and incorrectly to USFM2—please fix)

**Comment by PhotoNomad0** [comment]
After discussion with @RobH123 @lrsallee and @jag3773, we making a change that we do not put a new line before the alignment markers unless there is a space preceding.

**Comment by PhotoNomad0** [comment]
Larry's project: 
[en_ult_gal_book.1.zip](https://github.com/unfoldingWord-dev/translationCore/files/2861881/en_ult_gal_book.1.zip)

**Comment by mannycolon** [comment]
Tested on c6de621 (v1.1.3)

**Comment by PhotoNomad0** [comment]
@cckozie - the changes made are listed in the definition of done above (merged in from a couple other issues that were combined).

Note that Larry's comment above about `\s5` and `\q` markers is covered in https://github.com/unfoldingWord-dev/translationCore/issues/5802 that is still waiting on the USFM expert to weigh in.

**Comment by RobH123** [comment]
In that gal book referenced above, the first \s5 marker should be AFTER the \c 1 marker. Oh, they all should be after the \c markers in the file, so:
\c 2
\s5 (ts)
\p
\v 1 verse text

**Comment by RobH123** [comment]
I can't find a file with \k markers in it to check if the \k-s\\* have matching \k-e\\* markers. (If not, I think they're not correct USFM3.)

If they are standalone milestones, they should be \k\\* (like \ts\\*) although it's unclear to me from the USFM3 documentation (http://ubsicap.github.io/usfm/master/milestones/index.html) if either of those are technically correct (and it seems that we are unable to get clarifications on the standard at the present time).

**Comment by cckozie** [comment]
Question for @benjore @RobH123 
Per the USFM3 spec for the \usfm marker; "USFM version specification for the file. 
Used to identify the USFM version which a USFM editor / processor will be required to support in order to manage all markup found within the file."

In v1.1.3 (c6de621) we are inserting a \usfm 3.0 marker into all USFM exported files regardless of whether or not they have any 3.0 markers in them. This would potentially indicate to an editor/processor that it needs to handle USFM3 markers when it really only needs to handle USFM2. Is that acceptable?

**Comment by RobH123** [comment]
@cckozie I wouldn't recommend it just yet. A USFM2 importer (like Paratext 8.0 or Bibledit) might reject loading it.

However, over time as USFM2 fades and USFM3 prevails (esp. once the forthcoming Paratext 8.1 is released which will handle USFM3.0), USFM2 importers will become obsolete/irrelevant and then it won't matter.

It seems to me that the main difference between what we are calling USFM2 and USFM3 exports, is really with or without alignment data, i.e., the content (and size -> downloading/loading time) rather than the format is the major difference. So I don't really see having the USFM3.0 indicator there as a problem either.

(How's that for an ambiguous reply.) :)

**Comment by cckozie** [comment]
1.1.3 (982b0d9) 
Regarding "correct output format of `\zaln-s` markers (adding the `\*` terminator)";
I wasn't sure of the exact meaning of the statement, so I performed a couple tests importing a slightly modified version of the original usfm3 file Larry attached above. Both tests ended up with corrupted data, although I'm sure that this issue was intended to fix the problems discovered.

In the first test I removed the "\*" after the first \zaln-e in verse 1. I then imported the modified file and opened it in wA. Verse 1 now only contains the word Paul.

In the second test I removed the first "\zaln-e\*" in verse 1. I then imported the modified file and loaded it in wA. Verse 1 now has all Greek words joined and all English words aligned to the group.

Imported and exported files here: 
[test files.zip](https://github.com/unfoldingWord-dev/translationCore/files/2920618/test.files.zip)

----------------------------
Regarding "correct output format of `\k-s` markers (adding the `\*` terminator)":
I don't understand how we would know where to insert the '\\*' terminator, so I don't know how to test this.

**Comment by PhotoNomad0** [comment]
@cckozie - here is what is wrong with our current alignment format that does not match spec.  The red circles are flagging newlines that should not be there according to the spec (new line is used as a substitute for spaces within a verse, so extra newlines would be interpreted as extra spaces).  The blue circle indicates where there should be a terminator (`\*`) for the `\zaln-s` marker.

![screen shot 2019-03-04 at 11 56 50 am](https://user-images.githubusercontent.com/14238574/53751619-aa8fe500-3e7a-11e9-8163-9da85f014530.png)

**Comment by PhotoNomad0** [comment]
This is how it should look:
```
\v 1 \zaln-s | x-strong="G39720" x-lemma="Παῦλος" x-morph="Gr,N,,,,,NMS," x-occurrence="1" x-occurrences="1" x-content="Παῦλος"\*\w Paul|x-occurrence="1" x-occurrences="1"\w*\zaln-e\*,
\zaln-s | x-strong="G06520" x-lemma="ἀπόστολος" x-morph="Gr,N,,,,,NMS," x-occurrence="1" x-occurrences="1" x-content="ἀπόστολος"\*\w an|x-occurrence="1" x-occurrences="1"\w*
\w apostle|x-occurrence="1" x-occurrences="1"\w*\zaln-e\*—\zaln-s | x-strong="G37560" x-lemma="οὐ" x-morph="Gr,D,,,,,,,,," x-occurrence="1" x-occurrences="1" x-content="οὐκ"\*\w not|x-occurrence="1" x-occurrences="1"\w*\zaln-e\*
\zaln-s | x-strong="G05750" x-lemma="ἀπό" x-morph="Gr,P,,,,,G,,," x-occurrence="1" x-occurrences="1" x-content="ἀπ’"\*\w from|x-occurrence="1" x-occurrences="2"\w*\zaln-e\*
```

Basically the `\*` is supposed to indicate the end of the parameters for the `\zaln-s` marker.

**Comment by PhotoNomad0** [comment]
@cckozie @RobH123 - here is an example from our catalog with the `\k` markers https://git.door43.org/Door43-Catalog/el-x-koine_ugnt/src/branch/master/49-GAL.usfm.  Basically they are used to connect more that one word to a translationWord.  In the file (line 23) there is missing the `\*` at the end of the `\k-s` marker:

```
\k-s | x-tw="rc://*/tw/dict/bible/kt/jesus"
```

should be 
```
\k-s | x-tw="rc://*/tw/dict/bible/kt/jesus"\*
```

there are also extraneous newlines in the example file.

**Comment by PhotoNomad0** [comment]
@RobH123 @benjore @cckozie - the `\usfm 3.0` marker needs clarity.  Maybe it was thought that we would only add it to usfm export if there are alignments included.  If that was the case, it could cause some problems.  Here are some use cases to consider:

- if the user has edited and added usfm3 markers (there are a lot of them) in the project, do we add the `\usfm 3.0` marker on export to usfm without alignments?  To match the usfm spec we should probably add it.  But that means additional processing on export as there are a lot of them - we will need to create another list.
- if usfm content is imported that already has the `\usfm 3.0` marker, do we remove it if they export to usfm **without** alignments?  Do we also search the text to see if there are usfm3 markers before removing?

if we are just looking to have a flag in the exported usfm file to determine if there are alignments, we should probably use a different marker than `\usfm 3.0`, because that does not equal alignments in the usfm3 spec.

Currently I am just adding the `\usfm 3.0` marker on every USFM export (with or without alignments) since it seemed is would be the safer approach without adding a lot of processing.  But it sounds like from @RobH123 that most editors do not recognize this yet.

**Comment by RobH123** [comment]
Yeah, my test on Paratext 8.0 (which is not the forthcoming USFM3 compatible 8.1) shows that it simply ignores the \usfm 3.0 marker, in order words, it reads the file fine.

USFM2 will soon be forgotten except as a subset of USFM3 which I expect most Bible translation software will handle by the end of the year. So @PhotoNomad0 , I think the easiest solution to your problems above is just to always include it on the output/export.

**Comment by cckozie** [comment]
1.1.3 (5a14a95)
Regarding adding the \usfm 3.0 marker to usfm2 exported files, it appears to be acceptable at this point, so I'm writing up a new issue for it. (#5864 )
Regarding the comments about tests done on 1.1.3 (982b0d9):
The tests performed tested problems with the imported file which are not related to this issue. I am writing up another issue to deal with the problems of the missing \zaln-e* and missing * after the \zaln. (#5865 )
